dozer/cogs/roles.py

```diff
@@ -34,13 +36,75 @@ def calculate_epoch_time(time_string):
         time_release = round(time.time() + time_delta)
         return time_release
 
+    @staticmethod
+    async def safe_message_fetch(ctx, menu=None, channel=None, message_id=None):
+        """Used to safely get a message and raise an error message cannot be found"""
+        try:
+            if menu:
+                channel = ctx.guild.get_channel(menu.channel_id)
+                return await channel.fetch_message(menu.message_id)
+            else:
+                if channel:
+                    return await channel.fetch_message(message_id)
+                else:
+                    return await ctx.message.channel.fetch_message(message_id)
+        except discord.HTTPException:
+            raise BadArgument("That message does not exist or is not in this channel!")
+
+    @staticmethod
+    async def add_to_message(message, entry):
+        """Adds a reaction role to a message"""
+        await message.add_reaction(entry.reaction)
+        await entry.update_or_add()
+        return
+
+    @staticmethod
+    async def del_from_message(message, entry):
+        """Removes a reaction from a message"""
+        await message.clear_reaction(entry.reaction)
+
     @Cog.listener('on_ready')
     async def on_ready(self):
         """Restore tempRole timers on bot startup"""
         q = await TempRoleTimerRecords.get_by()  # no filters: all
         for record in q:
             self.bot.loop.create_task(self.removal_timer(record))
 
+    @Cog.listener()
+    async def on_raw_message_delete(self, payload):
+        """Used to remove dead reaction role entries"""
+        message_id = payload.message_id
+        await ReactionRole.delete(message_id=message_id)
+        await RoleMenu.delete(message_id=message_id)
+
+    @Cog.listener()
+    async def on_raw_reaction_add(self, payload):
+        """Raw API event for reaction add, passes event to action handler"""
+        await self.on_raw_reaction_action(payload)
+
+    @Cog.listener()
+    async def on_raw_reaction_remove(self, payload):
+        """Raw API event for reaction remove, passes event to action handler"""
+        await self.on_raw_reaction_action(payload)
+
+    async def on_raw_reaction_action(self, payload):
+        """Called whenever a reaction is added or removed"""
+        message_id = payload.message_id
+        reaction = str(payload.emoji)
+        reaction_roles = await ReactionRole.get_by(message_id=message_id, reaction=reaction)
```

Comment:
This might need to be cached with `db.ConfigCache`, since reactions are added and removed frequently.